netket/graph/lattice.py

```diff
@@ -64,99 +63,19 @@ def __repr__(self):
         return f"LatticeSite({s})"
 
 
-def create_sites(
-    basis_vectors, extent, apositions, pbc, order
-) -> Tuple[Sequence[LatticeSite], Sequence[bool], Dict[HashableArray, int]]:
-    # note: by modifying these, the number of shells can be tuned.
-    shell_vec = _np.where(pbc, 2 * order, 0)
-    shift_vec = _np.where(pbc, order, 0)
-
-    shell_min = 0 - shift_vec
-    shell_max = _np.asarray(extent) + shell_vec - shift_vec
-    # cell coordinates
-    ranges = [slice(lo, hi) for lo, hi in zip(shell_min, shell_max)]
-    # site coordinate within unit cell
-    ranges += [slice(0, len(apositions))]
-
-    basis_coords = _np.vstack([_np.ravel(x) for x in _np.mgrid[ranges]]).T
-    site_coords = (
-        basis_coords[:, :-1]
-        + _np.tile(apositions.T, reps=len(basis_coords) // len(apositions)).T
-    )
-    positions = site_coords @ basis_vectors
-
-    sites = []
-    coord_to_site = {}
-    for idx, (coord, pos) in enumerate(zip(basis_coords, positions)):
-        sites.append(
-            LatticeSite(
-                id=None,  # to be set later, after sorting all sites
-                basis_coord=coord,
-                position=pos,
-            ),
-        )
-        coord_to_site[HashableArray(coord)] = idx
-    is_inside = ~(
-        _np.any(basis_coords[:, :-1] < 0, axis=1)
-        | _np.any(basis_coords[:, :-1] > (extent - 1), axis=1)
-    )
-    return sites, is_inside, coord_to_site
+def _create_sites(basis_vectors, extent, site_offsets):
+    ranges = [slice(0, ex) for ex in [*extent, len(site_offsets)]]
+    basis_coords = _np.mgrid[ranges].reshape(len(extent) + 1, -1).T
+    positions = basis_coords[:, :-1] @ basis_vectors
+    positions = positions.reshape(-1, len(site_offsets), len(extent)) + site_offsets
+    positions = positions.reshape(-1, len(extent))
```

Comment:
These several lines of code for creating site positions in real-space and lattice coordinates exists now in both `lattice._create_sites` and `_lattice_edge_logic.create_padded_sites` in equivalent form (except for the later version also supporting extra padding sites).

I think it would be great to extract it into a small helper method to remove this duplication.